### PR TITLE
GH-474 Suggestion tooltip

### PR DESCRIPTION
**Pull request by huanmeng-qwq**
The suggested content includes tooltip text

### CODE REVIEW

**Comment by huanmeng-qwq** [comment]
img_v3_02fv_4e928ddb-af49-4c3b-a620-67d7e7c5f8dg

**Comment by huanmeng-qwq** [comment]
Ok.

**Comment by huanmeng-qwq** [comment]
@Rollczi 
Hi, can you review this PR?

**Comment by Rollczi** [comment]
I will review that on Monday

File: litecommands-core/src/dev/rollczi/litecommands/suggestion/Completion.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package dev.rollczi.litecommands.suggestion;
+
+import java.util.Objects;
+
+public class Completion {
```
Comment (Rollczi) [inline litecommands-core/src/dev/rollczi/litecommands/suggestion/Completion.java:5]: I think this class is useless and represents the same thing as Suggestion.

File: litecommands-unit/src/dev/rollczi/litecommands/unit/AssertSuggest.java [hunk 2]
```diff
@@ -51,6 +54,14 @@ public AssertSuggest assertSuggest(String... suggestions) {
         return assertSuggest(Arrays.asList(suggestions));
     }
 
+    public AssertSuggest assertSuggest(String[] suggestions, String... tooltips) {
```
Comment (Rollczi) [inline litecommands-unit/src/dev/rollczi/litecommands/unit/AssertSuggest.java:57]: If we want to assert suggestions with tooltip, then I think we should create method with varargs Suggestion... parameter

**Review by Rollczi — CHANGES_REQUESTED** [review]

**Comment by huanmeng-qwq** [comment]
I have completed the modifications. However, should we use tooltips on all platforms? Currently, I only apply tooltips to parameters registered on the fabric platform.

File: litecommands-core/src/dev/rollczi/litecommands/suggestion/SuggestionResult.java [hunk 3]
```diff
@@ -140,6 +151,20 @@ public static SuggestionResult of(Iterable<String> suggestions) {
         return new SuggestionResult(parsedSuggestions);
     }
 
+    public static SuggestionResult tooltip(Iterable<String> suggestions, Iterable<String> tooltips) {
```
Comment (Rollczi) [inline litecommands-core/src/dev/rollczi/litecommands/suggestion/SuggestionResult.java:154]: remove this method

File: litecommands-core/src/dev/rollczi/litecommands/suggestion/SuggestionResult.java [hunk 4]
```diff
@@ -155,7 +180,11 @@ public static SuggestionResult from(Iterable<Suggestion> suggestions) {
     }
 
     public static SuggestionResultCollector collector() {
-        return new SuggestionResultCollector();
+        return new SuggestionResultCollector(false);
+    }
+
+    public static SuggestionResultCollector tooltipCollector() {
```
Comment (Rollczi) [inline litecommands-core/src/dev/rollczi/litecommands/suggestion/SuggestionResult.java:186]: such tooltips don't make much sense

File: litecommands-core/src/dev/rollczi/litecommands/suggestion/SuggestionResultCollector.java [hunk 5]
```diff
@@ -9,6 +9,11 @@
 import java.util.stream.Collector;
 
 public class SuggestionResultCollector implements Collector<String, SuggestionResult, SuggestionResult> {
+    private final boolean copyTooltip;
+
+    public SuggestionResultCollector(boolean copyTooltip) {
```
Comment (Rollczi) [inline litecommands-core/src/dev/rollczi/litecommands/suggestion/SuggestionResultCollector.java:14]: rollback

File: litecommands-unit/src/dev/rollczi/litecommands/unit/AssertSuggest.java [hunk 6]
```diff
@@ -98,6 +98,15 @@ public AssertSuggest assertSuggest(Collection<String> suggestions) {
         return this;
     }
 
+    public AssertSuggest assertTooltip(Collection<String> tooltips) {
```
Comment (Rollczi) [inline litecommands-unit/src/dev/rollczi/litecommands/unit/AssertSuggest.java:101]: > If we want to assert suggestions with tooltip, then I think we should create method with varargs Suggestion... parameter

```suggestion
    public AssertSuggest assertSuggest(Suggestion... suggestions) {
```

File: litecommands-bukkit/src/dev/rollczi/litecommands/bukkit/TabCompleteProtocolLibAsync.java [hunk 7]
```diff
@@ -63,8 +64,7 @@ private void tryReplaceConsoleTabCompleter(Server server) {
     private Object getMinecraftServer(Object craftServer) {
         try {
             return ReflectUtil.getFromField(craftServer, "server");
-        }
-        catch (LiteCommandsReflectException exception) {
+        } catch (LiteCommandsReflectException exception) {
```
Comment (Rollczi) [inline litecommands-bukkit/src/dev/rollczi/litecommands/bukkit/TabCompleteProtocolLibAsync.java:67]: please don't change the code style for try catch

**Comment by Rollczi** [comment]
> I have completed the modifications. However, should we use tooltips on all platforms? Currently, I only apply tooltips to parameters registered on the fabric platform.

It would be useful to support all platforms that provide such possibilities (I guess e.g. JDA doesn't allow it)

**Comment by huanmeng-qwq** [comment]
@Rollczi Please review

**Comment by Rollczi** [comment]
Hi @huanmeng-qwq,
I added collector and copying of tooltip to new suggestions. Now tooltip is not null.

One important think: we need to think about tooltips like as experimental and unstable feature becasue it is only related to minecraft's domain and we need also to think about supporting adventure API. Thats means in the future I will need to change this API to support also components/json.

**Comment by huanmeng-qwq** [comment]
Yes, plain text only is currently supported, and future versions should support the use of the adventure API or the component api available on minecraft Vanilla on some platforms



---Original---
From: "Norbert ***@***.***&gt;
Date: Wed, Oct 30, 2024 07:01 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [Rollczi/LiteCommands] Suggestion tooltip (PR #461)




 
Hi @huanmeng-qwq,
 I added collector and copying of tooltip to new suggestions. Now tooltip is not null.
 
One important think: we need to think about tooltips like as experimental and unstable feature becasue it is only related to minecraft's domain and we need also to think about supporting adventure API. Thats means in the future I will need to change this API to support also components/json.

**Review by Rollczi — APPROVED** [review]

**Comment by huanmeng-qwq** [comment]
When can we merge this PR? or how should I use this snapshot version? @Rollczi

**Comment by Rollczi** [comment]
If you think all is done then we can merge it

**Comment by huanmeng-qwq** [comment]
> If you think all is done then we can merge it如果你认为一切都完成了那么我们可以合并它

Yes, it can be merged now

**Pull request merged**